rollout/trafficrouting.go

```diff
@@ -17,152 +17,177 @@ import (
-		if !atDesiredReplicaCount {
-			// Use the previous weight since the new RS is not ready for a new weight
-			for i := *index - 1; i >= 0; i-- {
-				step := c.rollout.Spec.Strategy.Canary.Steps[i]
-				if step.SetWeight != nil {
-					desiredWeight = *step.SetWeight
-					break
+		currentStep, index := replicasetutil.GetCurrentCanaryStep(c.rollout)
+		desiredWeight := int32(0)
+		weightDestinations := make([]trafficrouting.WeightDestination, 0)
+		if rolloututil.IsFullyPromoted(c.rollout) {
+			// when we are fully promoted. desired canary weight should be 0
+		} else if c.pauseContext.IsAborted() {
+			// when promote aborted. desired canary weight should be 0
+		} else if c.newRS == nil || c.newRS.Status.AvailableReplicas == 0 {
+			// when newRS is not available or replicas num is 0. never weight to canary
+		} else if index != nil {
+			atDesiredReplicaCount := replicasetutil.AtDesiredReplicaCountsForCanary(c.rollout, c.newRS, c.stableRS, c.otherRSs)
+			if !atDesiredReplicaCount {
+				// Use the previous weight since the new RS is not ready for a new weight
+				for i := *index - 1; i >= 0; i-- {
+					step := c.rollout.Spec.Strategy.Canary.Steps[i]
+					if step.SetWeight != nil {
+						desiredWeight = *step.SetWeight
+						break
+					}
 				}
+			} else if *index != int32(len(c.rollout.Spec.Strategy.Canary.Steps)) {
+				// This if statement prevents the desiredWeight from being set to 100
+				// when the rollout has progressed through all the steps. The rollout
+				// should send all traffic to the stable service by using a weight of
+				// 0. If the rollout is progressing through the steps, the desired
+				// weight of the traffic routing service should be at the value of the
+				// last setWeight step, which is set by GetCurrentSetWeight.
+				desiredWeight = replicasetutil.GetCurrentSetWeight(c.rollout)
 			}
-		} else if *index != int32(len(c.rollout.Spec.Strategy.Canary.Steps)) {
-			// This if statement prevents the desiredWeight from being set to 100
-			// when the rollout has progressed through all the steps. The rollout
-			// should send all traffic to the stable service by using a weight of
-			// 0. If the rollout is progressing through the steps, the desired
-			// weight of the traffic routing service should be at the value of the
-			// last setWeight step, which is set by GetCurrentSetWeight.
-			desiredWeight = replicasetutil.GetCurrentSetWeight(c.rollout)
-		}
 
-		// Checks for experiment step
-		// If current experiment exists, then create WeightDestinations for each experiment template
-		exStep := replicasetutil.GetCurrentExperimentStep(c.rollout)
-		if exStep != nil && c.currentEx != nil && c.currentEx.Status.Phase == v1alpha1.AnalysisPhaseRunning {
-			getTemplateWeight := func(name string) *int32 {
-				for _, tmpl := range exStep.Templates {
-					if tmpl.Name == name {
-						return tmpl.Weight
+			// Checks for experiment step
```

Comment:
same here

Reply:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right